extension/api.js

```diff
@@ -78,149 +105,257 @@ const lostEngines = new Map ([
+        async respond(allow) {
+          console.log(`reset-default-search: user responded to panel with allow? ${allow}`);
+          if (allow) {
+            await ExtensionSettingsStore.initialize();
+            ExtensionSettingsStore.addSetting(
+              extension.id,
+              DEFAULT_SEARCH_STORE_TYPE,
+              DEFAULT_SEARCH_SETTING_NAME,
+              engineName,
+              () => defaultEngine.name
+            );
+            Services.search.defaultEngine = Services.search.getEngineByName(
+              engineName
+            );
+          }
+
+          // Remember that we have completed.
+          finish(true, "interaction", allow ? "accepted" : "denied", extension.id);
+        },
+      },
+    };
+    Services.obs.notifyObservers(
+      subject,
+      "webextension-defaultsearch-prompt"
+    );
+
+    try {
+      Services.telemetry.recordEvent(
+        "defaultSearchReset",
+        "interaction",
+        "panelShown",
+        extension.id
+      );
+    } catch (err) {
+      // If the telemetry throws just log the error so it doesn't break any
+      // functionality.
+      Cu.reportError(err);
+    }
+
+    // We only prompt for the first addon that makes it here.  If the user does
+    // not respond to the panel, they will be asked again on next startup.
+    return;
+  }
+  // If we made it here, no addon was currently eligible.  If any addons were
+  // blocklisted we will want to be able to prompt when it is released from
+  // the blocklist.
+  tryagain = tryagain || enabledAddons.length < addons.length;
+  if (!tryagain) {
+    console.log("reset-default-search: no addons fit the criteria, finished.");
+    finish(false, "skipped", "noAddonsEligible");
+  } else {
+    console.log("reset-default-search: no enabled addons fit the criteria, waiting for more updates.");
+    try {
+      Services.telemetry.recordEvent(
+        "defaultSearchReset",
+        "interaction",
+        "tryLater",
+        null
+      );
+    } catch (err) {
+      // If the telemetry throws just log the error so it doesn't break any
+      // functionality.
+      Cu.reportError(err);
+    }
+  }
+}
+
+let _running = false;
+function runDefaultReset() {
+  if (_running) {
+    return;
+  }
+  _running = true;
+  tryDefaultReset()
+    .then(() => {
+      if (!tryagain) {
+        Management.off("ready", tryUpdate);
+      }
+    }).finally(() => {
+      _running = false;
+    });
+}
+
+function tryUpdate(e, { id }) {
+  if (lostEngines.has(id)) {
+    runDefaultReset();
+  }
+}
+
 this.search = class extends ExtensionAPI {
-  async onStartup() {
+  onStartup() {
+    console.log("reset-default-search: starting.");
+
     // We only run this once, after which we bail out.
     if (Services.prefs.prefHasUserValue(RUN_ONCE_PREF)) {
+      console.log("reset-default-search: has already ran once, exit.");
       return;
     }
+
     Services.telemetry.registerEvents("defaultSearchReset", {
+      skipped: {
```

Comment:
Can you also update data.md accordingly to cover the telemetry events and parameters we're adding?